consensus/ballot_state.go

```diff
@@ -529,18 +530,38 @@ func (s *BallotState) RelevantRange(x SlotValue) (int, int) {
 	min, max := 0, 0
 	for _, message := range s.M {
 		a, b := message.RelevantRange(x)
-		min, max = RangeUnion(min, max, a, b) 
+		min, max = RangeUnion(min, max, a, b)
 	}
 	return min, max
 }
 
+// Returns the max ballot number that a blocking set of nodes are talking about.
+func (s *BallotState) MaxActionableBallotNumber() int {
+	numberToNodes := make(map[int][]string)
+
+	for node, message := range s.M {
+		maxN := message.MaxN()
+		numberToNodes[maxN] = append(numberToNodes[maxN], node)
+	}
+
+	max := 0
+
+	for n, nodes := range numberToNodes {
```

Comment:
This makes it seem like if you have 3 other nodes and their maxes are 2, 3, and 4, then max will just be 0. That doesn't seem right? It seems like you need to be accumulating or something so that the nodes with a max=4 also count as making n=3 actionable.

Reply:
Yeah, that doesn’t seem right. I’ll take a look in a bit.

Reply:
This might not cause trouble with a full cluster but then cause trouble when there's a knockout because then it's harder to get a blocking set to agree on things. I would try running the knockout tests if you haven't already to see if you can find a deadlock.

Reply:
(with a higher-than-in-the-code # of loops i mean)